src/common/Locator.c

```diff
@@ -0,0 +1,230 @@
+#include <stdio.h>
+
+// Legacy ARDOPC used to transmit "No GS" in IDFRAMEs to indicate
+// that the grid square is unset. This is *not* valid as a
+// Packed6 since it was actually encoded with a lowercase "o."
+//
+// To support this, we quietly accept any of the following
+// byte sequences as an unset grid square.
+const static Packed6 LOCATOR_NOGS[] = {
+	{{0xbc, 0xf0, 0x27, 0xcc, 0x00, 0x00}},
+	{{0xba, 0xf0, 0x27, 0xcc, 0x00, 0x00}},
+};
+
+/**
+ * Validate locator
+ *
+ * Validate the `grid` field of the given `locator`.
+ *
+ * Returns LOCATOR_OK if the locator is valid or some
+ * other value if it is not.
+ */
+ARDOP_MUSTUSE locator_err
+locator_validate_grid(const Locator* locator);
+
+// true if s is ASCII alphabetical with the given
+// ending character `end`. For example, set `end = 'R'`
+// to accept alphabet characters A through R. `end` must
+// be uppercase.
+//
+// NOTE: ctype.h functions respect the current locale, and
+// we do *not* want that since it might not be an English
+// locale
+inline static bool is_ascii_alpha_range(const char s, const char end) {
+	const char LOWERCASE_ADD = 'a' - 'A';
+
+	bool upper = s >= 'A' && s <= end;
+	bool lower = s >= ('A' + LOWERCASE_ADD) && s <= (end + LOWERCASE_ADD);
+	return (upper || lower);
+}
+
+// convert given character to ASCII lowercase
+//
+// NOTE: ctype.h tolower() respects the current locale, and
+// we do *not* want that since it might not be an English
+// locale
+inline static char to_ascii_lowercase(const char s) {
+	const char LOWERCASE_ADD = 'a' - 'A';
+
+	if (s >= 'A' && s <= 'Z') {
+		return (char)(s + LOWERCASE_ADD);
+	} else {
+		return s;
+	}
+}
+
+// true if s is ASCII digit
+inline static bool is_ascii_digit(const char s) {
+	return s >= '0' && s <= '9';
+}
+
+// compress the grid square
+// the `grid` field must be populated
+static locator_err locator_compress(const Locator* locator, Packed6* out) {
+	locator_err res = locator_validate_grid(locator);
+	if (res) {
+		return res;
+	}
+
+	if (!packed6_from_str_slice(locator->grid, sizeof(locator->grid) - 1, out)) {
+		return LOCATOR_ERR_FMT_CHARSET;
+	}
+
+	return LOCATOR_OK;
+}
+
+// uncompress `grid` square from a Packed6
+// the wire bytes must be populated
+static locator_err locator_uncompress(const Packed6* in, Locator* locator) {
+	// check for legacy "unset gridsquare" byte sequence
+	for (size_t ind = 0; ind < sizeof(LOCATOR_NOGS)/sizeof(LOCATOR_NOGS[0]); ++ind) {
+		if (0 == memcmp(in, &LOCATOR_NOGS[ind], sizeof(*in))) {
+			locator_init(locator);
+			return LOCATOR_OK;
+		}
+	}
+
+	packed6_to_fixed_str(in, locator->grid);
+
+	// translate second letter pair to lowercase
+	for (size_t i = 4; i < 6; ++i) {
+		locator->grid[i] = to_ascii_lowercase(locator->grid[i]);
+	}
+
+	// translate space to NUL, which will truncate the grid square
+	for (size_t i = 0; i < sizeof(locator->grid); ++i) {
+		if (locator->grid[i] == ' ') {
+			locator->grid[i] = 0;
+		}
+	}
+
+	return locator_validate_grid(locator);
```

Comment:
This is a change in behavior.  Previously, received (compressed) gridsquare was not validated.  Whatever was received was passed to the host.  I don't believe that hosts actually do anything other than perhaps log/display the string.  So, I guess it doesn't cause any problem.